image-upload

```diff
@@ -28,46 +27,42 @@ import urllib.parse
 
 from lib.constants import BOTS_DIR, IMAGES_DIR
 from lib.directories import get_images_data_dir
-from lib.stores import S3_STORES, PUBLIC_STORES, REDHAT_STORES, CA_PEM
+from lib.stores import PUBLIC_STORES, REDHAT_STORES, CA_PEM
 from lib import s3
 
-from task import api
 
-
-def upload(dest, source, public, prune=False):
+def upload(store, source, public, prune=False):
+    dest = urllib.parse.urljoin(store, os.path.basename(source))
     url = urllib.parse.urlparse(dest)
 
+    if not s3.is_key_present(url):
+        # No credentials?  That's not going to work...
+        sys.stderr.write(f"image-upload: no credentials for {dest}")
+        return False
+
     # Start building the command
-    cmd = ["curl", "--no-progress-meter", "--fail", "--upload-file", source]
+    curl = ["curl", "--fail", "--upload-file", source]
 
-    # Pass credentials, if present
-    if s3.is_key_present(url):
-        headers = {s3.ACL: s3.PUBLIC} if public else {}
+    if store in REDHAT_STORES:
+        curl += ['--cacert', CA_PEM]
 
-        # slightly magic: we already know the checksum: it's part of the filename
-        hash_value = re.search('[0-9a-f]{64}', source).group(0)
+    headers = {s3.ACL: s3.PUBLIC} if public else {}
 
-        cmd += s3.sign_curl(url, method='PUT', headers=headers, checksum=hash_value)
+    # slightly magic: we already know the checksum: it's part of the filename
+    hash_value = re.search('[0-9a-f]{64}', source).group(0)
 
-    elif api.token:
-        user = url.username or getpass.getuser()
-        cmd += ["--user", user + ":" + api.token, '--cacert', CA_PEM, dest]
-
-    else:
-        # No credentials?  That's not going to work...
-        sys.stderr.write(f"image-upload: no credentials for {dest}")
-        return False
+    curl_arg = s3.sign_curl(url, method='PUT', headers=headers, checksum=hash_value)
 
     # Only prune after the above checks, to ensure that we have credentials.
     if prune:
-        subprocess.check_call([f'{BOTS_DIR}/image-prune', '--s3', f'https://{url.hostname}/'])
+        subprocess.check_call([f'{BOTS_DIR}/image-prune', '--s3', store])
```

Comment:
It does, I tested image pruning explicitly and also through `image-upload --s3-prune`. See the change in lib/s3.py, that does the same "apply ca.pem for REDHAT_STORES" approach as image-upload/download.

Reply:
I missed this, thanks.

In the name of DRY and keeping design decisions hidden within a single module, I think it would be better to also do one of:

1) add the `curl --cacert` logic to the curl helper in `lib.s3`, possibly moving over to always using the "new" signing for curl requests, and dropping the old code

2) port everything to use the `s3.urlopen()` path.  I mostly kept using `curl` because I thought we cared about the progress bar, but since you got rid of that, we don't really have a reason anymore.

Reply:
I separated and cleaned up the CA handling from the hosting type now.